### DISCUSSION TITLE
Migrating Nhost Functions to Node.js 18

### DISCUSSION THREAD

**Discussion by dbarrosop** [nhost/nhost#2239] [Changelog]
Dear Nhost Users,

We wanted to inform you about an upcoming change in our serverless functions service. As of October 1st, we will be replacing Node.js 16 with Node.js 18, the current LTS version.

Node.js 16 reached its end-of-life (EOL) on September 11th, and to ensure continued performance, security, and compatibility, we have decided to upgrade to Node.js 18. This upgrade will ensure continued support, security patches and stability.

To assist you with a smooth transition, we are introducing a new flag to the Nhost CLI. After upgrading to 1.6.0 you can make use of the ﻿flag `--node18` to run your code with Node.js 18 instead of the default Node.js 16. This will help you test and migrate your functions to the new runtime environment without any disruptions.

This migration to Node.js 18 will be done automatically on your behalf on your first deployment after October 1st.

We recommend reviewing your code and dependencies to ensure compatibility with Node.js 18 and testing with our CLI. If you encounter any challenges or have questions, our support team is here to assist you.

Thank you for your understanding and continued support.

Best regards,

The Nhost Team

---

### Quick CLI demo

Using the example `examples/myproject` on this repo:

1. Default behavior

```
$ nhost up
... (output omitted for brevity)


$ curl -s https://local.functions.nhost.run/v1/echo | jq
{
  "headers": {
    "host": "local.functions.nhost.run",
    "user-agent": "curl/8.1.2",
    "accept": "*/*",
    "x-forwarded-for": "192.168.64.1",
    "x-forwarded-host": "local.functions.nhost.run",
    "x-forwarded-port": "443",
    "x-forwarded-proto": "https",
    "x-forwarded-server": "1f7dd21f1fa2",
    "x-real-ip": "192.168.64.1",
    "x-replaced-path": "/v1/echo",
    "accept-encoding": "gzip"
  },
  "query": {},
  "node": "v16.20.0" <-------
}

2. Using `--node18` flag:

$ nhost up --node18
... (output omitted for brevity)

$ curl -s https://local.functions.nhost.run/v1/echo | jq
{
  "headers": {
    "host": "local.functions.nhost.run",
    "user-agent": "curl/8.1.2",
    "accept": "*/*",
    "x-forwarded-for": "192.168.64.1",
    "x-forwarded-host": "local.functions.nhost.run",
    "x-forwarded-port": "443",
    "x-forwarded-proto": "https",
    "x-forwarded-server": "1f7dd21f1fa2",
    "x-real-ip": "192.168.64.1",
    "x-replaced-path": "/v1/echo",
    "accept-encoding": "gzip"
  },
  "query": {},
  "node": "v18.17.1" <-------
}

```

**Comment by quentin-decre** [comment]
I would like to deploy with node 18 on a given Subdomain ? How is it possible ?

**Reply by dbarrosop** [reply]
It is not possible. You can use the CLI for testing.

**Comment by fungilation** [comment]
instead of the `--node18` flag. Should cli on local and on deployed envs, both not use nhost.toml config? There's this I have there:

```
[functions.node]
version = 16
```

**Reply by quentin-decre** [reply]
That would be great indeed !

**Reply by dbarrosop** [reply]
We thought about that but as it will not be respected in the cloud we decided against it. When we switch the cloud to node18 on October 1st we will require that value to be changed to `version = 18` breaking your deployments until it is fixed. This will serve both as a signal that the change is about to happen in the cloud and a blocker so users can block their production deployments until they have evaluated them in their staging environments (if that's what they desire to do). I will write a short guide next week describing what will happen on October 1st and what will need to be done. This post was mostly to get the CLI out and to let people know how to start testing their functions with node18.

I understand this isn't ideal and that people would rather control the runtime they use on their own (and we'd love to allow you to do that too) but, truth be told, that'd require us to do significant changes to how things work internally at a time when we are still trying to decide the future of functions (i.e. move from our current solution based on AWS lambda to some other based on edge functions)

**Reply by fungilation** [reply]
Have you considered not to unilaterally change everyone's functions node version, so `functions.node.version` is actually respected in cloud deployments?

Upgrading to 18 should have been an option earlier, I've asked for it. So I don't care about moving to it. But not having a transitional period and not giving users choice in version to deploy with isn't a good process, especially when there's config for it.

**Reply by dbarrosop** [reply]
> Have you considered not to unilaterally change everyone's functions node version, so functions.node.version is actually respected in cloud deployments?

Yes, but as I mentioned on my previous message:

> I understand this isn't ideal and that people would rather control the runtime they use on their own (and we'd love to allow you to do that too) but, truth be told, that'd require us to do significant changes to how things work internally at a time when we are still trying to decide the future of functions (i.e. move from our current solution based on AWS lambda to some other based on edge functions)
